### ISSUE TITLE
Dropdown list for profile attributes

### ISSUE THREAD

**Issue by Julien-Dubois-eXo** [Meeds-io/MIPs#171]
## Rationale

Profiles help users share useful information with others, such as contact details, managers, position, department, and more.

Profiles are now more accessible and functional thanks to these MIPs:
- Quick Search ([MIP](https://github.com/Meeds-io/MIPs/issues/110)): Profile attributes can be used for fast searches.
- Advanced Filtering ([MIP](https://github.com/Meeds-io/MIPs/issues/39)): Profile attributes enable filtering in the people directory.
These enhancements make it easier for users to find and connect with others.

To further improve profiles, admins have requested the ability to:
- Let users add contact details or profile attributes using dropdown lists.
- Manage these lists by adding, editing, or removing values.

#### List of prefilled elements
**Administration - Add list of prefield elements for an attribute**
**Given** I'm a platform/administrator
**When** I create or edit an attribute 
**Then** I have an option "Dropdown List" I can activate or deactivate

Once activated, I can edit a list of elements to add choices the users can select in their user profile: 
I can: 
- add an element
- translate the element by clicking on the translation option to open the standard translation drawer
- edit an existing element to correct it. The modification will be applied to the users who have selected the element.
- delete an element

If I deactivate the option, the list values will be kept. Once reactivated I still have them. 

![Image](https://github.com/user-attachments/assets/c7348a41-9bac-4bc0-8c99-30653766c88e)
![Image](https://github.com/user-attachments/assets/e00d0c54-1e06-46db-be05-9e69c4e2d8ed)

**Precisions:**
- This dropdown list option can only be used when the attribute type is a text (not user)
- Any option proposed for profile attribute (multivalues, required, visible, etc.) can be combined with this dropdown list option
- When the attribute was selected from a dropdown list, and this list is no longer activated (by admin), then we keep the value the user has selected. Once an edition of the profile is done, the attribute will be removed
- When the attribute was added without using a dropdown list, and the admin decides to provide a dropdown list, then we keep the value the user has manually added. Once an edition of the profile is done, the attribute will be removed
- When a value is deleted on the dropdown list by the administrator, then the user with this value selected will no longer display the attribute on their profile until they edit it to choose another value. 

**User - Select a choice on a prefilled attribute**
**Given** I'm a user on my profile
**When** I edit my user information in my profile
**Then** for the attribute with prefilled settings, I have a dropdown selector to choose from the list of elements the administrator defines. 
I can filter the dropdown list by typing a word in the input box as it's possible in the target selection dropdown in the publication drawer.

![Image](https://github.com/user-attachments/assets/d772066f-5f32-4fa9-b87c-a59715043215)

#### User attributes on analytics
**Administration - Attribute listed on the analytics**
**Given** I'm a platform/administrator
**When** I create or edit an attribute on the profile page. 
**Then** I can activate /deactivate the option "Indexed on the analytics"

![Image](https://github.com/user-attachments/assets/8f1f9aa4-43f4-4c07-a102-466b02fe4030)

Once activated the attribute will be automatically added to every operation with the attribute Social: User ID.

#### Analytics
**Analytics - User attribute on an operation**
**Given** I'm a platform/analytics admin or platform/admin or any user with access to analytics
**When** I have an operation with the attribute Social: User ID **Then** I have automatically another attribute "Social: [name of the profile attribute]" with its value for each profile attribute respecting the following:

- activated by the admin 
- filled
- not hidden by the user or the admin

I can:
- filter on the profile attribute and select a specific value to create a chart. 
- see the information in the "sample"

### Security 

Administration of values of a dropdown for a profile property is available only for administrator : creation, activation/deactivation, update, deletion

Once configured, the Profile properties must be read by all users and externals.

### Upgradability

Currently, ProfileSettings translations was done before the creation of TranslationAPI which use metadata.
It results in the usage of a specific database table (SOC_LABELS) and a specific UI. 

During the realization of this feature, we will try to align the usage of translation for propertySettings with the standard, and then remove the specific UI and table.
Then existing translations for profilepropertysetting will have to be migrated. 
If this point is done, we will add an upgrade plugin.

### Feature Flags

| Property Name  | Default Value | Target Audience | Functional Behaviour |
| ------------- | ------------- | ------------- | ------------- |
|  |   |   |   |

### Access 
   - GUI 
A new drawer will be added in profile property configuration
Profile UI (display and edit) will be updated to render and to allow update of a select option
   - API 
ProfileSettingsRest will be extended to use ProfilePropertyOptions
ProfilePropertyService will be extended to read and store the new possibility of PropertyPropertyOption

### Services & processing
   - Indexation : as the value of the dropdown must be used in the user search, we need to index for the user profile the translated values of the choosen option. We will index all possible translations, as we currently do for a multivalued property.
   - If the admin modifies a translation of an option, we need to reindex user profile which have choose this option, to take care of the new translation. This reindexation will be launched asynchronously.

### Data and persistence
   - Macro data model
![Profile Property Select Data Model](https://github.com/user-attachments/assets/dfecea1d-57a6-4ed4-a34b-84d792c6ad70)

The part in black already exists, the part in red is added.

   - In the user profile, we have to store the value (single or multiple), of the selected option. 
For that we will store the PROPERTY_OPTION_ID, in the profile field.
With that, if the admin change the translation of the option, we have nothing to update in the profile. 
At profile display, we convert the option_id in the translated string using the locale of the user which consult the profile.
When a user edit his profile, it will be easier to find which option is selected, using the option_id (the select html component will take display the option like this) : 
`<option value="42">Human Resouces</option>`

  - When an admin disable the dropdown option for a profile property, we do not remove the options. We only dont use it not more. So that, when admin reactivate it, we can reuse it.

  - When an admin removes an option in the list of options of a profile property, the profil of user which have choosen this option will consider the property as it is not filled.

## 5. Annexes
   - Reasoning that explain the technical choices
   - Impacts 
   - References

**Comment by Julien-Dubois-eXo** [comment]
@margondicco could you, please, check if I didn't forget something before I start the designs? @rdenarie FYI

**Comment by Julien-Dubois-eXo** [comment]
@srenault-meeds ready to review for gofonc please.

**Comment by srenault-meeds** [comment]
LGTM. Yet, I think two things are missing:

**Multivalues + dropdown list**
- Add the fact that multivalues option can be used for dropdown list 

**Behaviour if dropdown is not used or used after a while**

_First case: the attribute was used previously and dropdow list wasn't provided yet_
- Take an attribute that is not proposed with a validation list (dropdown list): team
- After some time, admins decide to provide for this attribute a dropdown list
->> How do you manage previously added attributes?
1. Do you remove informations previously added by users?
2. Do you keep it?

_Second case: the attribute was used previously, a dropdown was suggested for users. Admins decide to remove this list_
- Take an attribute filled thanks to a dropdown list
- After some time, admins decide to deactivate this option
->> How do you manage previously added attributes?
1. Do you remove information previously selected from the dropdown list by users?
2. Do you keep it?

**Deactivation of list**
- And last but not least, let's say you have provided a dropdown list for an attribute.
- You decide to deactivate it
- And few minutes, days, months later, you decide to activate it.
->> Do you provide previous values of the dropdown list?
1. If yes, that means we must store it somewhere
2. If yes, that means we must keep it during a limited duration (30 days for example)
3. If no, that means we need a confirmation message

Available to discuss it of course

**Comment by srenault-meeds** [comment]
After discussing it together, one only feedback to check with your team

> Deactivation of list

And last but not least, let's say you have provided a dropdown list for an attribute.
You decide to deactivate it
And few minutes, days, months later, you decide to activate it.
->> Do you provide previous values of the dropdown list?
If yes, that means we must store it somewhere
If yes, that means we must keep it during a limited duration (30 days for example)
If no, that means we need a confirmation message
Available to discuss it of course

**Comment by Julien-Dubois-eXo** [comment]
@srenault-meeds, I updated the description to keep the data in case the option is deactivated. Ready to final review for gofonc.

**Comment by srenault-meeds** [comment]
Ok. It might be useful to study it once this case to avoid cumulating lots of unuseful data.

No blocker for me. Go Fonc

**Comment by rdenarie** [comment]
@boubaker ready for review for tech part.

**Comment by boubaker** [comment]
ok for me

**Comment by rdenarie** [comment]
Hi @srenault-meeds @boubaker 

Acc mips and exo-mips, and PRs are ready for review

Thanks for your feedback

**Comment by boubaker** [comment]
> Hi [@srenault-meeds](https://github.com/srenault-meeds) [@boubaker](https://github.com/boubaker)
> 
> Acc mips and exo-mips, and PRs are ready for review
> 
> Thanks for your feedback

Thanks. as said before, unfortunately, we broke the Data Model Design by adding Analytics behavior inside Social Data Model. It would be nicer to have a dedicated Admin UI for Analytics where the administrators can choose User Attribute to index (A first setting type for a new Admin UI for Analytics where we will put other configuration stuff, such as permissions to set analytics charts inside spaces...)
But as discussed, we 're in a hurry, thus ok for me and no objection for merge.

**Comment by rdenarie** [comment]
I had trace your feedback about Data Model Design point.

**Comment by boubaker** [comment]
> Hi [@srenault-meeds](https://github.com/srenault-meeds) [@boubaker](https://github.com/boubaker)
> 
> Acc mips and exo-mips, and PRs are ready for review
> 
> Thanks for your feedback

Can you please change the labels of the MIP

**Comment by margondicco** [comment]
Hello 

feedbacks

https://github.com/Meeds-io/meeds/issues/2851

**Comment by srenault-meeds** [comment]
Hi,

My feedbacks:
- when switching on the option to provide to an attribute a dropdown list:
   - then I can proceed without adding any value
   - Problem: the user is invited to choose from a list, but nothing is provided
   - This is is bad UX IMO
<img width="389" alt="Image" src="https://github.com/user-attachments/assets/a481d74f-61f0-4553-badb-c455e4440f94" />

<img width="302" alt="Image" src="https://github.com/user-attachments/assets/b3a093a7-94e0-453c-86c1-31e7211db150" />

> **Suggestion**: when switching the option to True, open the second-level drawer so the admin can start adding values. This will reduce the fact that admins miss adding values for the dropdown list

- placeholder for choosing a value is not ok with UI guidelines. We must use a greycolorlighten1 var
<img width="305" alt="Image" src="https://github.com/user-attachments/assets/2dda580d-28a8-4773-8948-2fabee3b1085" />

- icon for removing the value chosen is not ok with UI guidelines. Please, DO NOT USE mdi icons...
<img width="283" alt="Image" src="https://github.com/user-attachments/assets/27f0fc52-404f-42ea-b81c-11ae9144eb2f" />

- UX problem when adding data or translation
   - I have added a translation (or I have added data in the list)
   - I have saved it
   - Then I got back to the second-level drawer with data list
      - Result: I thought it was ok but no, I forgot to save. 
      - Of course, a snackbar tells me it has not been saved but this is wrong UX and I have lost my data (feedback already reported months ago about another feature). Please, DO NOT USE IT. It doesn't follow guidelines explained [here](https://builders.meeds.io/portal/g/:spaces:developers/home/apps/notes/315)

<img width="267" alt="Image" src="https://github.com/user-attachments/assets/06549809-84f0-411c-b4fd-28539f9c689d" />

![Image](https://github.com/user-attachments/assets/1b2decf1-d037-4dbd-8697-d4f7e54b166d)

> **Suggestion**: when clicking outside the drawer without saving the changes, inform the user edits will not be saved

**Comment by Julien-Dubois-eXo** [comment]
@srenault-meeds my answers: 

We will treat these 3 feedbacks: 
> Suggestion: when switching the option to True, open the second-level drawer so the admin can start adding values. This will reduce the fact that admins miss adding values for the dropdown list

> placeholder for choosing a value is not ok with UI guidelines. We must use a greycolorlighten1 var

> icon for removing the value chosen is not ok with UI guidelines. Please, DO NOT USE mdi icons...



The last feedback will not be treated here as it was already implemented before, it was not part of the scope evaluated initially in the MIP. I have reported it as an improvement and it will be treated as a quick win by the maintenance team. https://github.com/Meeds-io/meeds/issues/2861

**Comment by Julien-Dubois-eXo** [comment]
@srenault-meeds we have implemented the fix about the 3 feedbacks. You can test them. We have also fixed Morgan's feedback https://github.com/Meeds-io/meeds/issues/2851. 

Let us know if when it's ok for you to merge the MIP.

**Comment by srenault-meeds** [comment]
Thanks @Julien-Dubois-eXo.

It is ok for me regarding these issues.
It is indeed ok.

However, a new behaviour to fix IMO.
- if an attribute has values for a dropdown list
- When editing it, it opens the second-level drawer to add more values
See GIF below

Problem: I didn't want to edit these values but I wanted to edit the attribute itself.

Thanks.

![Image](https://github.com/user-attachments/assets/d5711ac5-9b2d-43c7-8cd4-74456753950e)

**Comment by Julien-Dubois-eXo** [comment]
@srenault-meeds, we have just fixed this point. I have tested it. Can you validate from your side too please?

**Comment by srenault-meeds** [comment]
Ok for me. Thanks.
Validated
